File: ios/Plugin/AccountCommon.swift
/*
 * AccountCommon Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in the root directory.
 */

import Foundation

/// A Swift struct representing common properties of an account, such as its type and source.
public struct AccountCommon{
    
    /// The type of the account, represented by an AccountTypeEnum.
    let type: AccountTypeEnum
    /// The source of the account, typically a string identifier.
    let source: String
    
    /// A dictionary of default AccountCommon objects for various sources.
    static let defaults: [String: AccountCommon] = [
        "ACME_MARKETS" : AccountCommon(type: .retailer, source: RetailerEnum.ACME_MARKETS.rawValue),
        "ALBERTSONS" : AccountCommon(type: .retailer, source: RetailerEnum.ALBERTSONS.rawValue),
        "AMAZON": AccountCommon(type: .retailer, source: RetailerEnum.AMAZON_BETA.rawValue),
        "AMAZON_CA": AccountCommon(type: .retailer, source: RetailerEnum.AMAZON_CA.rawValue),
        "AMAZON_UK": AccountCommon(type: .retailer, source: RetailerEnum.AMAZON_UK.rawValue),
        "BED_BATH_AND_BEYOND": AccountCommon(type: .retailer, source: RetailerEnum.BED_BATH_AND_BEYOND.rawValue),
        "BESTBUY": AccountCommon(type: .retailer, source: RetailerEnum.BESTBUY.rawValue),
        "BJS_WHOLESALE": AccountCommon(type: .retailer, source: RetailerEnum.BJS_WHOLESALE.rawValue),
        "CHEWY": AccountCommon(type: .retailer, source: RetailerEnum.CHEWY.rawValue),
        "COSTCO": AccountCommon(type: .retailer, source: RetailerEnum.COSTCO.rawValue),
        "CVS": AccountCommon(type: .retailer, source: RetailerEnum.CVS.rawValue),
        "DICKS_SPORTING_GOODS": AccountCommon(type: .retailer, source: RetailerEnum.DICKS_SPORTING_GOODS.rawValue),
        "DOLLAR_GENERAL": AccountCommon(type: .retailer, source: RetailerEnum.DOLLAR_GENERAL.rawValue),
        "DOLLAR_TREE": AccountCommon(type: .retailer, source: RetailerEnum.DOLLAR_TREE.rawValue),
        "DOMINOS_PIZZA": AccountCommon(type: .retailer, source: RetailerEnum.DOMINOS_PIZZA.rawValue),
        "DOOR_DASH": AccountCommon(type: .retailer, source: RetailerEnum.DOOR_DASH.rawValue),
        "DRIZLY": AccountCommon(type: .retailer, source: RetailerEnum.DRIZLY.rawValue),
        "FAMILY_DOLLAR": AccountCommon(type: .retailer, source: RetailerEnum.FAMILY_DOLLAR.rawValue),
        "FOOD_4_LESS": AccountCommon(type: .retailer, source: RetailerEnum.FOOD_4_LESS.rawValue),
        "FOOD_LION": AccountCommon(type: .retailer, source: RetailerEnum.FOOD_LION.rawValue),
        "FRED_MEYER": AccountCommon(type: .retailer, source: RetailerEnum.FRED_MEYER.rawValue),
        "GAP": AccountCommon(type: .retailer, source: RetailerEnum.GAP.rawValue),
        "GIANT_EAGLE": AccountCommon(type: .retailer, source: RetailerEnum.GIANT_EAGLE.rawValue),
        "GRUBHUB": AccountCommon(type: .retailer, source: RetailerEnum.GRUBHUB.rawValue),
        "HEB": AccountCommon(type: .retailer, source: RetailerEnum.HEB.rawValue),
        "HOME_DEPOT": AccountCommon(type: .retailer, source: RetailerEnum.HOME_DEPOT.rawValue),
        "HYVEE": AccountCommon(type: .retailer, source: RetailerEnum.HYVEE.rawValue),
        "INSTACART": AccountCommon(type: .retailer, source: RetailerEnum.INSTACART.rawValue),
        "JEWEL_OSCO": AccountCommon(type: .retailer, source: RetailerEnum.JEWEL_OSCO.rawValue),
        "KOHLS": AccountCommon(type: .retailer, source: RetailerEnum.KOHLS.rawValue),
        "KROGER": AccountCommon(type: .retailer, source: RetailerEnum.KROGER.rawValue),
        "LOWES": AccountCommon(type: .retailer, source: RetailerEnum.LOWES.rawValue),
        "MACYS": AccountCommon(type: .retailer, source: RetailerEnum.MACYS.rawValue),
        "MARSHALLS": AccountCommon(type: .retailer, source: RetailerEnum.MARSHALLS.rawValue),
        "MEIJER": AccountCommon(type: .retailer, source: RetailerEnum.MEIJER.rawValue),
        "NIKE": AccountCommon(type: .retailer, source: RetailerEnum.NIKE.rawValue),
        "PUBLIX": AccountCommon(type: .retailer, source: RetailerEnum.PUBLIX.rawValue),
        "RALPHS": AccountCommon(type: .retailer, source: RetailerEnum.RALPHS.rawValue),
        "RITE_AID": AccountCommon(type: .retailer, source: RetailerEnum.RITE_AID.rawValue),
        "SAFEWAY": AccountCommon(type: .retailer, source: RetailerEnum.SAFEWAY.rawValue),
        "SAMS_CLUB": AccountCommon(type: .retailer, source: RetailerEnum.SAMS_CLUB.rawValue),
        "SEAMLESS": AccountCommon(type: .retailer, source: RetailerEnum.SEAMLESS.rawValue),
        "SEPHORA": AccountCommon(type: .retailer, source: RetailerEnum.SEPHORA.rawValue),
        "SHIPT": AccountCommon(type: .retailer, source: RetailerEnum.SHIPT.rawValue),
        "SHOPRITE": AccountCommon(type: .retailer, source: RetailerEnum.SHOPRITE.rawValue),
        "SPROUTS": AccountCommon(type: .retailer, source: RetailerEnum.SPROUTS.rawValue),
        "STAPLES": AccountCommon(type: .retailer, source: RetailerEnum.STAPLES.rawValue),
        "STAPLES_CA": AccountCommon(type: .retailer, source: RetailerEnum.STAPLES.rawValue),
        "STARBUCKS": AccountCommon(type: .retailer, source: RetailerEnum.STARBUCKS.rawValue),
        "TACO_BELL": AccountCommon(type: .retailer, source: RetailerEnum.TACO_BELL.rawValue),
        "TARGET": AccountCommon(type: .retailer, source: RetailerEnum.TARGET.rawValue),
        "TJ_MAXX": AccountCommon(type: .retailer, source: RetailerEnum.TJ_MAXX.rawValue),
        "UBER_EATS": AccountCommon(type: .retailer, source: RetailerEnum.UBER_EATS.rawValue),
        "ULTA": AccountCommon(type: .retailer, source: RetailerEnum.ULTA.rawValue),
        "VONS": AccountCommon(type: .retailer, source: RetailerEnum.VONS.rawValue),
        "WALGREENS": AccountCommon(type: .retailer, source: RetailerEnum.WALGREENS.rawValue),
        "WALMART": AccountCommon(type: .retailer, source: RetailerEnum.WALMART.rawValue),
        "WALMART_CA": AccountCommon(type: .retailer, source: RetailerEnum.WALMART_CA.rawValue),
        "WEGMANS": AccountCommon(type: .retailer, source: RetailerEnum.WEGMANS.rawValue),
        "YAHOO" : AccountCommon(type: .email, source: EmailEnum.YAHOO.rawValue),
        "OUTLOOK" : AccountCommon(type: .email, source: EmailEnum.OUTLOOK.rawValue),
        "AOL" : AccountCommon(type: .email, source: EmailEnum.AOL.rawValue),
        "GMAIL" : AccountCommon(type: .email, source: EmailEnum.GMAIL.rawValue),
        "GMAILIMAP" : AccountCommon(type: .email, source: EmailEnum.GMAIL.rawValue)
    ]
}

File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/AccountCommon.kt
package com.mytiki.sdk.capture.receipt.capacitor

/**
 * Enum of all account providers.
 *
 * This enum lists all the account providers supported by the system.
 *
 * @property type The type of the provider, which can be either Email or Retailer.
 * @property source The source of the provider.
 * @constructor Creates an empty [AccountCommon] enum entry.
 */
enum class AccountCommon(val type: AccountTypeEnum, val source: String) {
    ACME_MARKETS(AccountTypeEnum.RETAILER, RetailerEnum.ACME_MARKETS.toString()),
    ALBERTSONS(AccountTypeEnum.RETAILER, RetailerEnum.ALBERTSONS.toString()),
    AMAZON(AccountTypeEnum.RETAILER, RetailerEnum.AMAZON.toString()),
    AMAZON_CA(AccountTypeEnum.RETAILER, RetailerEnum.AMAZON_CA.toString()),
    AMAZON_UK(AccountTypeEnum.RETAILER, RetailerEnum.AMAZON_UK.toString()),
    BED_BATH_AND_BEYOND(AccountTypeEnum.RETAILER, RetailerEnum.BED_BATH_AND_BEYOND.toString()),
    BESTBUY(AccountTypeEnum.RETAILER, RetailerEnum.BESTBUY.toString()),
    BJS_WHOLESALE(AccountTypeEnum.RETAILER, RetailerEnum.BJS_WHOLESALE.toString()),
    CHEWY(AccountTypeEnum.RETAILER, RetailerEnum.CHEWY.toString()),
    COSTCO(AccountTypeEnum.RETAILER, RetailerEnum.COSTCO.toString()),
    CVS(AccountTypeEnum.RETAILER, RetailerEnum.CVS.toString()),
    DICKS_SPORTING_GOODS(AccountTypeEnum.RETAILER, RetailerEnum.DICKS_SPORTING_GOODS.toString()),
    DOLLAR_GENERAL(AccountTypeEnum.RETAILER, RetailerEnum.DOLLAR_GENERAL.toString()),
    DOLLAR_TREE(AccountTypeEnum.RETAILER, RetailerEnum.DOLLAR_TREE.toString()),
    DOMINOS_PIZZA(AccountTypeEnum.RETAILER, RetailerEnum.DOMINOS_PIZZA.toString()),
    DOOR_DASH(AccountTypeEnum.RETAILER, RetailerEnum.DOOR_DASH.toString()),
    DRIZLY(AccountTypeEnum.RETAILER, RetailerEnum.DRIZLY.toString()),
    FAMILY_DOLLAR(AccountTypeEnum.RETAILER, RetailerEnum.FAMILY_DOLLAR.toString()),
    FOOD_4_LESS(AccountTypeEnum.RETAILER, RetailerEnum.FOOD_4_LESS.toString()),
    FOOD_LION(AccountTypeEnum.RETAILER, RetailerEnum.FOOD_LION.toString()),
    FRED_MEYER(AccountTypeEnum.RETAILER, RetailerEnum.FRED_MEYER.toString()),
    GAP(AccountTypeEnum.RETAILER, RetailerEnum.GAP.toString()),
    GIANT_EAGLE(AccountTypeEnum.RETAILER, RetailerEnum.GIANT_EAGLE.toString()),
    GRUBHUB(AccountTypeEnum.RETAILER, RetailerEnum.GRUBHUB.toString()),
    HEB(AccountTypeEnum.RETAILER, RetailerEnum.HEB.toString()),
    HOME_DEPOT(AccountTypeEnum.RETAILER, RetailerEnum.HOME_DEPOT.toString()),
    HYVEE(AccountTypeEnum.RETAILER, RetailerEnum.HYVEE.toString()),
    INSTACART(AccountTypeEnum.RETAILER, RetailerEnum.INSTACART.toString()),
    JEWEL_OSCO(AccountTypeEnum.RETAILER, RetailerEnum.JEWEL_OSCO.toString()),
    KOHLS(AccountTypeEnum.RETAILER, RetailerEnum.KOHLS.toString()),
    KROGER(AccountTypeEnum.RETAILER, RetailerEnum.KROGER.toString()),
    LOWES(AccountTypeEnum.RETAILER, RetailerEnum.LOWES.toString()),
    MACYS(AccountTypeEnum.RETAILER, RetailerEnum.MACYS.toString()),
    MARSHALLS(AccountTypeEnum.RETAILER, RetailerEnum.MARSHALLS.toString()),
    MEIJER(AccountTypeEnum.RETAILER, RetailerEnum.MEIJER.toString()),
    NIKE(AccountTypeEnum.RETAILER, RetailerEnum.NIKE.toString()),
    PUBLIX(AccountTypeEnum.RETAILER, RetailerEnum.PUBLIX.toString()),
    RALPHS(AccountTypeEnum.RETAILER, RetailerEnum.RALPHS.toString()),
    RITE_AID(AccountTypeEnum.RETAILER, RetailerEnum.RITE_AID.toString()),
    SAFEWAY(AccountTypeEnum.RETAILER, RetailerEnum.SAFEWAY.toString()),
    SAMS_CLUB(AccountTypeEnum.RETAILER, RetailerEnum.SAMS_CLUB.toString()),
    SEAMLESS(AccountTypeEnum.RETAILER, RetailerEnum.SEAMLESS.toString()),
    SEPHORA(AccountTypeEnum.RETAILER, RetailerEnum.SEPHORA.toString()),
    SHIPT(AccountTypeEnum.RETAILER, RetailerEnum.SHIPT.toString()),
    SHOPRITE(AccountTypeEnum.RETAILER, RetailerEnum.SHOPRITE.toString()),
    SPROUTS(AccountTypeEnum.RETAILER, RetailerEnum.SPROUTS.toString()),
    STAPLES(AccountTypeEnum.RETAILER, RetailerEnum.STAPLES.toString()),
    STAPLES_CA(AccountTypeEnum.RETAILER, RetailerEnum.STAPLES_CA.toString()),
    STARBUCKS(AccountTypeEnum.RETAILER, RetailerEnum.STARBUCKS.toString()),
    TACO_BELL(AccountTypeEnum.RETAILER, RetailerEnum.TACO_BELL.toString()),
    TARGET(AccountTypeEnum.RETAILER, RetailerEnum.TARGET.toString()),
    TJ_MAXX(AccountTypeEnum.RETAILER, RetailerEnum.TJ_MAXX.toString()),
    UBER_EATS(AccountTypeEnum.RETAILER, RetailerEnum.UBER_EATS.toString()),
    ULTA(AccountTypeEnum.RETAILER, RetailerEnum.ULTA.toString()),
    VONS(AccountTypeEnum.RETAILER, RetailerEnum.VONS.toString()),
    WALGREENS(AccountTypeEnum.RETAILER, RetailerEnum.WALGREENS.toString()),
    WALMART(AccountTypeEnum.RETAILER, RetailerEnum.WALMART.toString()),
    WALMART_CA(AccountTypeEnum.RETAILER, RetailerEnum.WALMART_CA.toString()),
    WEGMANS(AccountTypeEnum.RETAILER, RetailerEnum.WEGMANS.toString()),
    YAHOO(AccountTypeEnum.EMAIL, EmailEnum.YAHOO.toString()),
    OUTLOOK(AccountTypeEnum.EMAIL, EmailEnum.OUTLOOK.toString()),
    AOL(AccountTypeEnum.EMAIL, EmailEnum.AOL.toString()),
    GMAIL(AccountTypeEnum.EMAIL, EmailEnum.GMAIL.toString());


    /**
     * Converts the [AccountCommon] object into a string representation.
     *
     * @return The name of the enum entry as a string.
     */
    override fun toString() = this.name

    companion object {
        /**
         * Finds and returns an [AccountCommon] enum entry with the specified [stringValue].
         *
         * @param stringValue The [AccountCommon.source] to search for.
         * @return The matching [AccountCommon] enum entry, or null if not found.
         */
        fun fromSource(source: String) = AccountCommon.values().first { it.source == source }

    }

}